### PR TITLE
Fix Pipelined Flaky Restart Recovery

### PR DESCRIPTION
**Pull request by koolzz**
Signed-off-by: Nick Yurchenko <koolzz@fb.com>

## Summary
 - Pull out restart logic into a separate mixin
 - Make inout controller rules persist on restarts(was causing momentary traffic disruptions)
 - Improve flow/flow_msg matching logic in message.py
 - Improve rule removal logic in restart_mixin.py

## Test Plan

 - Unit tests work fine

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

**Comment by 119Vik** [comment]
RestartCWAGTests

File: lte/gateway/python/magma/pipelined/app/restart_mixin.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+
+    def _remove_extra_flows(self, extra_flows):
+        msg_list = []
+        for tbl in extra_flows:
+            for flow in extra_flows[tbl]:
+                match = ovs_flow_match_to_magma_match(flow)
+                self.logger.debug('Sending msg for deletion -> %s',
+                                  match.ryu_match)
+                msg_list.append(flows.get_delete_flow_msg(
+                    self._datapath, self.tbl_num, match, cookie=flow.cookie,
```
Comment (pshelar) [inline lte/gateway/python/magma/pipelined/app/restart_mixin.py]: here the table number should be 'tbl'.

File: lte/gateway/python/magma/pipelined/app/restart_mixin.py [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+        except ControllerNotReadyException as err:
+            self.logger.error('Setup failed: %s', err)
+            return SetupFlowsResult(result=SetupFlowsResult.FAILURE)
+
+        self.logger.debug('Setting up %s default rules', self.APP_NAME)
+        remaining_flows = self._install_default_flows_if_not_installed(
+            self._datapath, startup_flows)
+
+        for tbl in startup_flows:
+            self.logger.debug('Startup flows before filstering -> %s',
```
Comment (pshelar) [inline lte/gateway/python/magma/pipelined/app/restart_mixin.py]: typo.

File: lte/gateway/python/magma/pipelined/app/inout.py [hunk 3]
```diff
@@ -144,51 +152,84 @@ def _get_config(self, config_dict):
...
 
     def initialize_on_connect(self, datapath):
-        self.delete_all_flows(datapath)
-        self._install_default_ingress_flows(datapath)
-        self._install_default_egress_flows(datapath)
-        self._install_default_middle_flows(datapath)
+        self._datapath = datapath
         self._setup_non_nat_monitoring(datapath)
 
+    def _install_default_flows_if_not_installed(self, datapath,
```
Comment (pshelar) [inline lte/gateway/python/magma/pipelined/app/inout.py:158]: functionality to generate flaw table diff can be moved to common lib in RestartMixin.

File: lte/gateway/python/magma/pipelined/app/restart_mixin.py [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+from magma.pipelined.policy_converters import FlowMatchError, \
+    convert_ipv4_str_to_ip_proto, ovs_flow_match_to_magma_match
+
+
+class RestartMixin(metaclass=ABCMeta):
+    """
+    RestartMixin
+
+    Mixin class for policy enforcement apps that includes common methods
+    used for rule activation/deactivation.
```
Comment (pshelar) [inline lte/gateway/python/magma/pipelined/app/restart_mixin.py:32]: It would be nice to extend this functionality to all controllers.

**Review by pshelar — APPROVED** [review]

File: lte/gateway/python/magma/pipelined/app/inout.py (hunk 3 above)
Comment (koolzz) [reply lte/gateway/python/magma/pipelined/app/inout.py:158]: Yeah good points I'll just keep the msg creation controller side and pull out all these into the restart mixin next diff

File: lte/gateway/python/magma/pipelined/app/restart_mixin.py (hunk 4 above)
Comment (koolzz) [reply lte/gateway/python/magma/pipelined/app/restart_mixin.py:32]: yup will work on that

**Pull request merged**